cmd/juju/charmhub/printinfo.go

```diff
@@ -0,0 +1,535 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package charmhub
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/juju/charm/v7"
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"golang.org/x/crypto/ssh/terminal"
+	"gopkg.in/yaml.v2"
+
+	"github.com/juju/juju/api/charmhub"
+	"github.com/juju/juju/cmd/output"
+)
+
+func printInfo(ctx *cmd.Context, in *charmhub.InfoResponse) error {
+	iw := newInfoWriter(ctx, in, termWidth(ctx.Stdout))
+	return iw.Print()
+}
+
+type InfoWriter interface {
+	Print() error
+}
+
+func newInfoWriter(ctx *cmd.Context, in *charmhub.InfoResponse, width int) InfoWriter {
+	iw := &infoWriter{
+		w:         ctx.Stdout,
+		warningf:  ctx.Warningf,
+		termWidth: width,
+		in:        in,
+	}
+	if iw.in.Type == "charm" {
+		return &charmInfoWriter{infoWriter: iw}
+	}
+	return &bundleInfoWriter{infoWriter: iw}
+}
+
+type infoWriter struct {
+	warningf  func(format string, params ...interface{})
+	w         io.Writer
+	termWidth int
+
+	in *charmhub.InfoResponse
+}
+
+func (iw *infoWriter) maybePrintName() error {
+	if iw.in.Name == "" {
+		return nil
+	}
+	_, err := fmt.Fprintf(iw.w, "name:        %s\n", iw.in.Name)
+	return err
+}
+
+func (iw *infoWriter) maybePrintSummary() error {
+	if iw.in.Charm.Summary == "" {
+		return nil
+	}
+	return wrapFlow(iw.w, quotedIfNeeded(iw.in.Charm.Summary), "summary:     ", iw.termWidth)
+}
+
+func (iw *infoWriter) maybePrintPublisher() error {
+	publisher, ok := iw.in.Charm.Publisher["display-name"]
+	if !ok {
+		return nil
+	}
+	_, err := fmt.Fprintf(iw.w, "publisher:   %s\n", publisher)
+	return err
+}
+
+func (iw *infoWriter) maybePrintDescription() error {
+	if iw.in.Charm.Description == "nil" {
+		return nil
+	}
+
+	if len(iw.in.Charm.Description) < iw.termWidth-13 {
+		_, err := fmt.Fprintf(iw.w, "description: %s\n", iw.in.Charm.Description)
+		return err
+	}
+
+	if _, err := fmt.Fprintln(iw.w, "description: |"); err != nil {
+		return err
+	}
+
+	return printDescr(iw.w, iw.in.Charm.Description, iw.termWidth)
+}
+
+func (iw *infoWriter) maybePrintStoreURL() error {
+	// TODO (hml) 2020-06-24
+	// implement once the data is available.
+	return nil
+}
+
+func (iw *infoWriter) maybePrintChannels() error {
+	// TODO (hml) 2020-06-24
+	// Implement up arrow for closed channels
```

Comment:
Do we have enough information to work that out?

Reply:
We do, but it's a bit ugly and will take some api/apiserver changes to do.  I've made a card: https://trello.com/c/4xjRg73L/2737-add-closed-channels-to-info-output